[4.x] Add passkey 🔑 support to CP logins

ryanmitchell opened the pull request:
This PR adds WebAuthn/Passkey support to the CP login:

https://github.com/statamic/cms/assets/51899/97eca681-2d55-478c-8ed8-0dfd88aa796d

We make use of the following libraries for backend and front end support respectively:
https://github.com/web-auth/webauthn-lib
https://simplewebauthn.dev/docs/packages/browser/

I noticed CraftCMS have used the same libraries for their Passkey support.

The user flow is as follows:

- An already authenticated user goes to the Passkeys page (new option under user menu) where they can create or remove Passkeys.
- Once a passkey has been created (through the Browsers in-built UI) the user then has the option to login using that on the CP login page. It should 'just work'.

The changes behind the scenes are:

- There is a new webauthn config file, allow this functionality to be enabled
- There is a new passkeys stache store (inside a new passkeys folder in the users folder)
- There is a passkeys repository and query builder to allow querying of a user's passkeys

I still need to add support for eloquent stored passkeys, and the UI on the passkeys page in the CP could do with being improved by someone who has better design skills than me.

robdekort commented:
This is amazing. I just did a test run and with SSL enabled locally and `"web-auth/webauthn-lib": "^4.7"` added to my `composer.json` I was able to create a passkey and login. Very hot stuff!

A side note regarding the login flow - and this should probably be addressed for oAuth as well in a separate PR - I'd ditch the Login with email button:

<img width="512" alt="Screenshot_2023-12-22_at_12 22 09" src="https://github.com/statamic/cms/assets/69107412/1080f09d-42da-4f79-8ece-2511128013cd">

And instead always show the regular login fields and below that options for other types of login: passkeys, oauth. This is basically the flow GitHub uses:

https://github.com/statamic/cms/assets/69107412/bfec98c0-cc70-4377-871b-ae6b7c64f439

robdekort commented:
One thing I noticed is that the users table lists the key as a user.
<img width="1584" alt="Screenshot 2023-12-22 at 12 35 44" src="https://github.com/statamic/cms/assets/69107412/f587a815-8a9f-4519-a071-c0acefef3ab3">

jasonvarga commented:
![](https://media.giphy.com/media/XreQmk7ETCak0/giphy.gif)

robdekort commented:
One note regarding this that I don't remember being in this PR. I feel Statamic should disable the option to login with password for the account that has a Passkey set. That seems to be the standard, as otherwise there's no security benefit, but it's just a convenience.

jasonvarga commented:
Looking at GitHub as an example, I have set up a passkey there, but they still give me the option to sign in with a username and password. Both work.

robdekort commented:
> Looking at GitHub as an example, I have set up a passkey there, but they still give me the option to sign in with a username and password. Both work.

I see, you’re right regarding GitHub. I’m honestly surprised by that. That is not what I see most sites that offer Passkeys do. For example my accounting software completely disables password logins as soon as you set a Passkey. And so does Sony on their recently introduced Passkeys for PS5. 

I believe this _should_ be how it’s done. Passkeys are here to eliminate passwords and improve security. By keeping the ability to login with a password there’s no security benefit gained.

jasonvarga commented:
Do you mean the option is gone from the form? Or you just get denied if you try to enter a username/password?

jasonvarga commented:
I have a PS account so I could test that myself. It seems that you have to enter your username first, then it gives you the appropriate option. e.g. I entered my email then it showed me the option to sign in with a passkey. I tried again with a fake email and it shows me the password field.

robdekort commented:
> Do you mean the option is gone from the form? Or you just get denied if you try to enter a username/password?

Yeah, getting denied. Sony disabled my password and 2fa, no way to enable it with a Passkey enabled. Here you can see it. Sorry Dutch, but I think you'll understand:
<img width="1043" alt="Screenshot 2024-04-18 at 21 02 18" src="https://github.com/statamic/cms/assets/69107412/205f0883-cf89-4811-b577-83819ab8edb4">

This is my accounting software. Also Dutch, but when I try to login with my e-mail it says nope:
<img width="547" alt="Screenshot 2024-04-18 at 21 03 54" src="https://github.com/statamic/cms/assets/69107412/dfb5feb3-c003-40bd-8b9c-2fdbda627e01">

robdekort commented:
Weirdly enough Apple does offer the password option. Perhaps because I had 2fa enabled, I'm not sure.
<img width="548" alt="Screenshot 2024-04-18 at 21 05 56" src="https://github.com/statamic/cms/assets/69107412/87c8cf97-0137-4ef8-9384-cb46ed05c8ee">

I guess it's a choice. But security wise I don't think it makes sense to keep the option.

jasonvarga commented:
Nice I love inconsistency. So I think maybe we do this:

Have an option that prevents logging in with passwords if you have a passkey set.

- Users without a passkey can still log in with email/password.
- Users _with_ a passkey that enter their password can still log in, if you configure it that way.
- Users _with_ a passkey that enter their password will be denied, if you configure it that way.

robdekort commented:
That sounds like a perfect solution. For example, I would enforce it by default. But maybe not if I’d use the 2fa addon. All options possible. I love it.

ryanmitchell commented:
Sounds good to me too. Let me know if you want me to update this to work that way.

ryanmitchell commented:
@jasonvarga I've updated this:

- It now targets 5.x
- Passkeys not longer have a store, querybuilder or repository. In stache they are stored in the user yaml, in eloquent we query the model directly
- I added some tests for adding and deleting passkeys (test coverage of the passkey functionality itself should come from the library not Statamic?)

ryanmitchell commented:
That sort of UI change is above my pay grade :)

edalzell commented:
> That sort of UI change is above my pay grade :)

I'll chip in £5

robdekort commented:
Ok, I think I've got all the various states now. Happy to make more changes. This is what it looks like:

**Normal login**
<img width="443" alt="Screenshot 2024-04-24 at 09 16 42" src="https://github.com/statamic/cms/assets/69107412/b3300722-372a-4182-8d96-9dc1426c0b89">

**Passkeys enabled**
<img width="428" alt="Screenshot 2024-04-24 at 09 16 52" src="https://github.com/statamic/cms/assets/69107412/070a0e21-9412-4571-af4e-dec6717adbca">

**Passkeys and oAuth enabled**
<img width="441" alt="Screenshot 2024-04-24 at 09 17 11" src="https://github.com/statamic/cms/assets/69107412/91d3df79-fa03-4b1c-90c8-3b1af701f02c">

**E-mail login disabled**
<img width="424" alt="Screenshot 2024-04-24 at 09 17 20" src="https://github.com/statamic/cms/assets/69107412/7a061770-d30f-4829-84ca-3a7aa918cfa1">

**Only oAuth enabled**
<img width="394" alt="Screenshot 2024-04-24 at 09 17 28" src="https://github.com/statamic/cms/assets/69107412/7239b208-5116-48ef-a6ca-0622eb6dfe2c">

jasonvarga commented:
I think this is done on my end. If anyone wants to give it a quick test, go for it. But we'll probably merge it tomorrow.

robdekort commented:
Very cool! Curious to test it out.

ryanmitchell commented:
Amazing. Thank you.

robdekort commented:
This works really well! A few things I've noticed, let me know if you'd rather have any of those as issues.

With `allow_password_login_with_passkey` set to false I was still able to login with my username and password. 

The JS alert to put in your passkey name might be more in line with the rest of the CP as a modal with a text input field.

jasonvarga commented:
Interesting. Seems fine for me.

<img width="367" height="219" alt="CleanShot 2025-10-25 at 15 55 27" src="https://github.com/user-attachments/assets/63371cc1-15f0-4e34-857c-b0db9d7e83b4" />

And yeah I agree about the JS prompts. Rushed that part a bit.

If you'd like to make separate issues that'd be good.
